searx/query.py

```diff
@@ -63,100 +70,225 @@ def _parse_query(self):
-
-                # check if prefix is equal with categorie name
-                elif prefix in categories:
-                    # using all engines for that search, which
-                    # are declared under that categorie name
-                    searx_query_part = True
-                    self.enginerefs.extend(EngineRef(engine.name, prefix)
-                                           for engine in categories[prefix]
-                                           if (engine.name, prefix) not in self.disabled_engines)
-
-            if query_part[0] == '!':
-                self.specific = True
+                # external bang
+                value = query_part[2:]
+                if len(value) > 0:
+                    searx_query_part = self._external_bang_parse(value, enable_autocomplete)
+                elif enable_autocomplete:
+                    self._external_bang_autocomplete()
+            elif query_part[0] == '!' or query_part[0] == '?':
+                # this force a engine or category
+                value = query_part[1:].replace('-', ' ').replace('_', ' ')
+                if len(value) > 0:
+                    searx_query_part = self._bang_parse(query_part[0], value, enable_autocomplete)
+                    if searx_query_part and query_part[0] == '!':
+                        self.specific = True
+                elif enable_autocomplete:
+                    self._bang_autocomplete(query_part[0], value)
 
             # append query part to query_part list
-            if searx_query_part:
-                self.query_parts.append(query_part)
+            qlist = self.query_parts if searx_query_part else self.user_query_parts
+
+            qlist.append(query_part)
+            last_index_location = (qlist, len(qlist) - 1)
+
+        self.autocomplete_location = last_index_location
+
+    def _timeout_autocomplete(self):
+        self.autocomplete_list.append('<3')
+        self.autocomplete_list.append('<850')
+
+    def _timeout_parse(self, raw_value):
+        try:
+            raw_timeout_limit = int(raw_value)
+        except ValueError:
+            # error not reported to the user
+            return False
+        else:
+            if raw_timeout_limit < 100:
+                # below 100, the unit is the second ( <3 = 3 seconds timeout )
+                self.timeout_limit = float(raw_timeout_limit)
             else:
-                self.user_query_parts.append(query_part)
+                # 100 or above, the unit is the millisecond ( <850 = 850 milliseconds timeout )
+                self.timeout_limit = raw_timeout_limit / 1000.0
+            return True
+
+    def _language_autocomplete(self, value):
+        if len(value) == 0:
+            # show some example queries
+            self.autocomplete_list.append(":en")
+            self.autocomplete_list.append(":en_us")
+            self.autocomplete_list.append(":english")
+            self.autocomplete_list.append(":united_kingdom")
+        else:
```

Comment:
Instead of putting these instructions in `else`, you could just return at the end of adding example quearies and we could win back one level of indentation.
```python
if len(value) == 0:
    for lang in [":en", ":en_us", ":english", ":united_kingdom"]:
        self.autocomplete_list.append(lang)
     return

for lc in language_codes:
    lang_id, lang_name, country, english_name = map(str.lower, lc)
    # etc
```

Reply:
Done